File: Demo/OC/test/ViewController.m
//
//  ViewController.m
//  test
//
//  Created by yanzhen on 2020/8/21.
//  Copyright © 2020 yanzhen. All rights reserved.
//

#import "ViewController.h"
#import <MSBPlayer/MSBPlayer.h>
#import "libyuv.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIImageView *smallPlayer;
@property (weak, nonatomic) IBOutlet UISlider *slider;
@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (nonatomic, strong) CIContext *context;

@property (nonatomic, strong) MSBAIPlayer *player;
@property (nonatomic, assign) BOOL isRead;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self playVideo];
    return;
    CVPixelBufferRef pixelBuffer = NULL;
    NSDictionary *pixelAttributes = @{(NSString *)kCVPixelBufferIOSurfacePropertiesKey:@{}};
    CVReturn result = CVPixelBufferCreate(kCFAllocatorDefault,
                                            928,
                                            1200,
                                            kCVPixelFormatType_32BGRA,
                                            (__bridge CFDictionaryRef)(pixelAttributes),
                                            &pixelBuffer);
    if (result != kCVReturnSuccess) {
        NSLog(@"MSB 002 Unable to create cvpixelbuffer %d", result);
        return;
    }
    
    NSLog(@"0000---%d:%d", CVPixelBufferGetBytesPerRow(pixelBuffer), CVPixelBufferGetHeight(pixelBuffer));
    
}

- (void)playVideo {
    NSString *path = [NSBundle.mainBundle pathForResource:@"3" ofType:@"mp4"];
    NSURL *pathUrl = [NSURL fileURLWithPath:path];
    
    NSURL *url = [NSURL URLWithString:@"http://39.107.116.40/res/tpl/default/file/guoke.mp4"];
    
    _player = [[MSBAIPlayer alloc] initWithURL:pathUrl mode:MSBVideoDecoderModeDisplayLayer];
    [_player attachToView:self.view];
    
    __weak ViewController *weakSelf = self;
    _player.playerStatus = ^(AVPlayerStatus status, NSError *error) {
        NSLog(@"11 playerStatus: %ld : %@", (long)status, error);
        if (status == AVPlayerStatusReadyToPlay) {
            [weakSelf.player play];
        }
    };
    
    
    _player.playbackStatus = ^(MSBAIPlaybackStatus status) {
        NSLog(@"22 playbackStatus: %ld", (long)status);
    };
    
    _player.playbackTime = ^(NSTimeInterval time, NSTimeInterval duration) {
        weakSelf.slider.maximumValue = duration;
        weakSelf.slider.value = time;
        weakSelf.timeLabel.text = [NSString stringWithFormat:@"%.f/%.f", time, duration];
    };
    
    _player.loadedTime = ^(NSTimeInterval time, NSTimeInterval duration) {
//        NSLog(@"44 loadedTime: %f/%f", time, duration);
    };
    
    
//    _player.audioDataBlock = ^(int sampleRate, int channels, void *data, int size) {
//        NSLog(@"55 audioDataBlock: %d, %d, %d", sampleRate, channels, size);
//    };
    
    
    _player.videoDataBlock = ^(CVPixelBufferRef pixelBuffer) {
        [weakSelf displayVideo:pixelBuffer];
    };
    
    
//    _player.yuvDataBlock = ^(int width, int height, NSData *data) {
//        @autoreleasepool {//must do this
//            [weakSelf autoBGRA:width height:height data:data];
//        }
//    };
    
}

- (void)autoBGRA:(int)width height:(int)height data:(NSData *)data {
    uint8_t *buffer = (uint8_t *)data.bytes;
    
    uint8_t *bgraBuffer = malloc(width * height * 4);
    I420ToARGB(buffer, width, buffer + width * height, width /  2, buffer + width * height * 5 / 4, width /  2, bgraBuffer, width * 4, width, height);

    CVPixelBufferRef pixelBuffer = NULL;
    CVReturn result = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, bgraBuffer, width * 4, NULL, NULL, NULL, &pixelBuffer);
    if (result != kCVReturnSuccess) {
        NSLog(@"MSB 002 Unable to create cvpixelbuffer %d", result);
        return;
    }
    
    [self displayVideo:pixelBuffer];
    CVPixelBufferRelease(pixelBuffer);
    free(bgraBuffer);
    bgraBuffer = NULL;
}

- (IBAction)playOrPause:(UIButton *)sender {
    if (sender.isSelected) {
        [_player resume];
    } else {
        [_player pause];
    }
    sender.selected = !sender.isSelected;
}

- (IBAction)sliderAction:(UISlider *)sender {
    [_player seekToTime:sender.value];
}

- (void)displayVideo:(CVPixelBufferRef)pixelBuffer {
    CIImage *ciImage = [CIImage imageWithCVImageBuffer:pixelBuffer];
    size_t width = CVPixelBufferGetWidth(pixelBuffer);
    size_t height = CVPixelBufferGetHeight(pixelBuffer);
    
    CGImageRef imageRef = [self.context createCGImage:ciImage fromRect:CGRectMake(0, 0, width, height)];
    if (!imageRef) {
        NSLog(@"Create Image fail with CVPixelBufferRef");
        return;
    }
    
    UIImage *image = [UIImage imageWithCGImage:imageRef];
    CGImageRelease(imageRef);
    dispatch_async(dispatch_get_main_queue(), ^{
        self.smallPlayer.image = image;
    });
    
}

-(CIContext *)context {
    if (!_context) {
        _context = [CIContext contextWithOptions:nil];
    }
    return _context;
}

@end

File: Demo/Swift/test/ViewController.swift
//
//  ViewController.swift
//  test
//
//  Created by Work on 2020/6/20.
//  Copyright © 2020 yanzhen. All rights reserved.
//

import UIKit
import MSBPlayer

private let video = "http://s2.meixiu.mobi/courseware/test/texiao/ES2U102KABWL/errors/2.mp4"




class ViewController: UIViewController {

    private var index = 1
    let context  = CIContext(options: nil)
    @IBOutlet private weak var smallPlayer: UIImageView!
    @IBOutlet private weak var slider: UISlider!
    @IBOutlet private weak var timeLabel: UILabel!
    private var player: MSBAIPlayer?
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        playVideo()
    }
    
    
    
    @IBAction private func seekTo(_ sender: UISlider) {
        player?.seek(toTime: TimeInterval(slider.value))
    }
    

    @IBAction private func pausePlay(_ sender: UIButton) {
        if sender.isSelected {
            player?.resume()
        } else {
            player?.pause()
        }
        sender.isSelected.toggle()
    }
    
}

private extension ViewController {
    func playVideo() {
        let path = Bundle.main.path(forResource: "1", ofType: "mp4")
        
        let url = URL(fileURLWithPath: path!)
//        let url = URL(string: "http://39.107.116.40/res/tpl/default/file/guoke.mp4")
            //URL(string: "https://xiaoxiong-private.oss-cn-hangzhou.aliyuncs.com/courseware/product/temp/0ac7c090c4d611ea8ad70b2b8db29493.mp4")
        
        
//        player = MSBAIApplePlayer(url: url)
//        player = MSBAIPlayer(url: url)
        player = MSBAIPlayer(url: url, mode: .toolBoxSync)
        player?.videoGravity = .resizeAspect
        player?.attach(to: view)
    
        
        player?.playerStatus = { [weak self] (status, error) in
            print("status:", status.rawValue, error)
            if status == .readyToPlay {
                self?.player?.play()
            }
        }
        
        player?.playbackStatus = { [weak self] (status) in
            guard let self = self else {
                return
            }
            print(status.rawValue, "end")
            if status == .ended {
//                self.player?.stop()
//                self.player = nil
//                self.index += 1
//                if self.index < 12 {
//                    self.playVideo()
//                }
            }
        }
                
        player?.playbackTime = { [weak self] (time, duration) in
            self?.slider.maximumValue = Float(duration)
            self?.slider.value = Float(time)
            self?.timeLabel.text = String(format: "%.f/%.f", time, duration)
        }
        
        player?.audioDataBlock = { (sampleRate, channels, data, size) in
            print(1234, sampleRate, channels, data, size);
        }
        
        player?.videoDataBlock = { [weak self] (buffer) in
            self?.displayVideo(buffer)
        }
        
        player?.loadedTime = { (time, duration) in
//            print("load:", time, duration)
        }
    }
}


private extension ViewController {
    func displayVideo(_ pixelBuffer: CVPixelBuffer?) {
        guard let pixelBuffer = pixelBuffer else { return }
        let ciImage = CIImage(cvImageBuffer: pixelBuffer)
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        guard let videoImageRef = context.createCGImage(ciImage, from: .init(x: 0, y: 0, width: width, height: height)) else {
            return
        }
        let image = UIImage(cgImage: videoImageRef)
        DispatchQueue.main.async {
            self.smallPlayer.image = image
        }
    }
}
